File: class2relational/src/main/java/org/atlanmod/model/dynamic/classModel/ClassClass.java
package org.atlanmod.model.dynamic.classModel;
import scala.collection.JavaConverters;

import java.util.ArrayList;
import java.util.List;

public class ClassClass extends ClassElement {

    public ClassClass(String id, String name){
        super(ClassMetamodel.CLASS);
        super.eSetProperty("id", id);
        super.eSetProperty("name", name);
    }

    public String getId(){ return (String) super.eGetProperty("id"); }

    public String getName(){ return (String) super.eGetProperty("name"); }

    @SuppressWarnings("unchecked")
    public List<ClassAttribute> getAttributes(){
        return ((List<ClassAttribute>) super.eGetReference("attributes"));
    }

    @SuppressWarnings("unchecked")
    public void addAttributes(List<ClassAttribute> attributes){
        super.eSetReference("attributes", new ArrayList<ClassAttribute>());
        ((List<ClassAttribute>) super.eGetReference("attributes")).addAll(attributes) ;
    }

    @SuppressWarnings("unchecked")
    public void addAttributes(scala.collection.immutable.List<ClassAttribute> attributes){
        super.eSetReference("attributes", new ArrayList<ClassAttribute>());
        ((List<ClassAttribute>) super.eGetReference("attributes")).addAll(JavaConverters.seqAsJavaList(attributes)) ;
    }

    @SuppressWarnings("unchecked")
    public void addAttribute(ClassAttribute attribute){
        super.eSetReference("attributes", new ArrayList<ClassAttribute>());
        ((List<ClassAttribute>) super.eGetReference("attributes")).add(attribute) ;
    }

    public void setId(String id){
        super.eSetProperty("id", id);
    }

    public void setName(String name){
        super.eSetProperty("name", name);
    }

    @Override
    public String toString() {
        return super.getType() +
                " (" +
                "id:" + this.getId() +
                ", name:" + this.getName() +
                ")";
    }
}

File: relational2class/src/main/scala/org/atlanmod/model/classmodel/ClassClass.scala
package org.atlanmod.model.classmodel

import org.atlanmod.model.IdGenerator
import org.atlanmod.tl.util.ListUtils

class ClassClass extends ClassElement(ClassMetamodel.CLASS) {

    @deprecated("Having a random ID can turn inconsistent the output of a transformation")
    def this(name: String) {
        this()
        val id: String = IdGenerator.id()
        super.eSetProperty("id", id)
        super.eSetProperty("name", name)
        super.eSetProperty("super", List(): List[String])
    }

    def this(id:String, name: String) {
        this()
        super.eSetProperty("id", id)
        super.eSetProperty("name", name)
        super.eSetProperty("super", List(): List[String])
    }

    @deprecated("Having a random ID can turn inconsistent the output of a transformation")
    def this(name: String, super_ : List[String]) = {
        this()
        val id: String = IdGenerator.id()
        super.eSetProperty("id", id)
        super.eSetProperty("name", name)
        super.eSetProperty("super", super_)
    }

    def this(id:String,name: String, super_ : List[String]) = {
        this()
        super.eSetProperty("id", id)
        super.eSetProperty("name", name)
        super.eSetProperty("super", super_)
    }

    override def getId: String = super.eGetProperty("id").asInstanceOf[String]
    def getName: String = super.eGetProperty("name").asInstanceOf[String]
    def getSuper: List[String] = super.eGetProperty("super").asInstanceOf[List[String]]
    def setName(name: String): Unit = super.eSetProperty("name", name)

    override def toString: String =
        getType + "([" + getId + (if (getSuper.nonEmpty) " <: " + getSuper.mkString(",") else "") + "] " + getName + ")"

    override def weak_equals(o: Any): Boolean = {
        o match {
            case obj: ClassClass =>
                this.getName.equals(obj.getName) && ListUtils.eqList(this.getSuper, obj.getSuper)
            case _ => false
        }
    }

}
